### DISCUSSION TITLE
SSR: Change graphql headers in dynamic routes nextjs

### DISCUSSION THREAD

**Discussion by rohanrajpal** [urql-graphql/urql#1237] [Q&A] [answered]
Hey, I want to achieve the following with graphql, nextjs and urql SSR

1. I have setup a dynamic route `[store-name].tsx`
2. Now the `store-name` will be the value of the header `token` in my graphql url
3. This token will now be used for any of the future graphql requests

So if the graphql url was `example.com/graphql`, it shall now be `example.com/graphql?token=store-name`

I have setup SSR according to https://formidable.com/open-source/urql/docs/advanced/server-side-rendering/#nextjs. So I'm not sure where can I add the code to receive the store name, and change the grahpql headers accordingly.

If someone knows how to go about it, would be great if you could help

**Comment by JoviDeCroock** [comment, marked as answer]
In your [store-name].tsx

```
function Component() {
  const router = useRouter()
  useQuery({
    query: YOUR_QUERY,
    context: React.useMemo(function () { return { url: `example.com/graphql?token=${router.query['store-name']}` } }, []) 
  })
}
```

**Reply by kitten** [reply]
On a side-note, I'm not sure why your API accepts an operational parameter as a query string parameter on what's typically a POST request, but I'd recommend looking into sending extra parameters via something other than with the URL

**Reply by rohanrajpal** [reply]
@JoviDeCroock Thanks a lot! That seems to do the work. I was wondering if there is a way to set the `token` header globally? Since then i would have to manually set the context every time for subsequent queries.

Was thinking if I could inject the header somewhere here
```ts

export default withUrqlClient((_ssrExchange, ctx) => ({
  // ...add your Client options here
  url: 'http://localhost:3000/graphql',
}))(Index);
```

Of course, the issue is that we can only access the Router inside the component to get the `store-name.` I think we should be able to get the `store-name` from the `ctx` variable as well? I cant seem to find any help for the same in their documentation.

**Reply by rohanrajpal** [reply]
> On a side-note, I'm not sure why your API accepts an operational parameter as a query string parameter on what's typically a POST request, but I'd recommend looking into sending extra parameters via something other than with the URL

Thanks for the reply. Yeah even I was wondering of a better way, basically the token I'm setting is from https://www.vendure.io/docs/storefront/shop-api-guide/ which is the `vendure-token` under the heading **Universal Parameters**

I guess specifying the token as a query parameter should be a better option? I was just preferring the header option since then I need not specify the token on every query.

**Reply by kitten** [reply]
You can use the `fetchOptions` config on the `Client` to pass either an object or a function returning an object that creates this header.

If this was for authentication we also have a specialised helper for that: https://formidable.com/open-source/urql/docs/advanced/authentication/

And there's also a way to alter `fetchOptions` by writing a custom exchange if the value is very dynamic https://formidable.com/open-source/urql/docs/common-questions/#how-do-we-achieve-asynchronous-fetchoptions

**Reply by rohanrajpal** [reply]
Hey @kitten, so the variable I need to pass to the header is `store-name` from `pages/[store-name].tsx`. However I cant figure out a way to use the router to query and obtain the `store-name` in the client config. As it is outside the react component.

Basically I want to do something like this

```ts
const router = useRouter()
export default withUrqlClient((_ssrExchange, ctx) => ({
  url: `https://example.com/graphql`,
  fetchOptions: {
    headers: {
      "token": router.query["store-name"],
    },
  },
}))(Post);
```
The issue with the above approach is that its illegal to create hooks like that. Do you happen to know how shall i go about it?

Like the ctx of [getserversideprops](https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering) has params which has the store name, does the ctx of `withUrqlClient` also have something similar?

**Reply by kitten** [reply]
Those are the same `ctx` arguments 🙇

**Reply by rohanrajpal** [reply]
But `ctx` comes out to be `undefined` if I do the following :worried:
```ts
export default withUrqlClient((_ssrExchange, ctx) => ({
  url: `https://example.com/graphql`,
  fetchOptions: {
    headers: {
      //"default-token" is getting sent
      "token": ctx ? ctx.params.id : "default-token",
    },
  },
}))(Post);
```
 Is it re-renderd and later context becomes defined?

**Reply by JoviDeCroock** [reply]
Context is only defined on your server-side-render on the client-side you'll have to use the static reference to your router which can be retrieved by doing: `import Router from 'next/router'`. There is nothing next-urql can do about this, this is a Next thing.

So

```
export default withUrqlClient((_ssrExchange, ctx) => ({
  url: `https://example.com/graphql`,
  fetchOptions: {
    headers: {
      //"default-token" is getting sent
      "token": ctx ? ctx.params.id : Router.query.id,
    },
  },
}))(Post);
```

**Reply by rohanrajpal** [reply]
Oh I see, I kind of get it now. I tried your method, but seems like the static Reference has no query parameter :(

![image](https://user-images.githubusercontent.com/7023147/102515082-4d970a80-40b3-11eb-8d83-b7acc011ad49.png)

As this is a Next issue now, you can mark the first reply as the answer. Thanks a lot for your help @JoviDeCroock @kitten ! :smile: Got to learn a lot of new things. 

But if you know a little on how to solve this, would be great if you can share.

**Reply by JoviDeCroock** [reply]
do a console.log on your Router, it's going to be on there but the place might have changed 😅 sorry not an expert myself

**Reply by rohanrajpal** [reply]
So on console log it returns a function (which is the constructor of Router I believe), so not sure if what I'm trying to achieve is even feasible or not.

**Reply by JoviDeCroock** [reply]
@rohanrajpal I didn't look at your code-fragment well, you have to return the fetchOptions object from a function, then the Router will work correctly.

```
  fetchOptions: () => ({
    headers: {
      //"default-token" is getting sent
      "token": ctx ? ctx.params.id : Router.query.id,
    },
  }),
```

**Reply by rohanrajpal** [reply]
On console log i get 
![image](https://user-images.githubusercontent.com/7023147/102517849-8edce980-40b6-11eb-9837-2dc01360bc4b.png)
Which on double clicking 
![image](https://user-images.githubusercontent.com/7023147/102517912-a025f600-40b6-11eb-8500-3bc94ba4e47f.png)
Which is the class definition of Router and seems like the only static variable is `events`

**Reply by rohanrajpal** [reply]
> @rohanrajpal I didn't look at your code-fragment well, you have to return the fetchOptions object from a function, then the Router will work correctly.
> 
> ```
>   fetchOptions: () => ({
>     headers: {
>       //"default-token" is getting sent
>       "token": ctx ? ctx.params.id : Router.query.id,
>     },
>   }),
> ```

Well I guess we can only use Router on the client side

![image](https://user-images.githubusercontent.com/7023147/102532119-7c1fe000-40c9-11eb-8af6-4a2180711048.png)

**Reply by JoviDeCroock** [reply]
ctx should always be defined on the server-side and undefined on the client-side so you're running into something weird, that being said I don't really think that's an urql concern 😅 

I'll check this in a codesandbox in a bit

**Reply by JoviDeCroock** [reply]
@rohanrajpal I made an example for you https://codesandbox.io/s/vigorous-bell-c8qdg?file=/pages/%5Bid%5D.js

**Reply by rohanrajpal** [reply]
Thanks a lot @JoviDeCroock ! Everything is working as expected now. You're a lifesaver.

Mistakes I was making
- I didnt pass {ssr:true} :facepalm: The docs [didnt have](https://formidable.com/open-source/urql/docs/advanced/server-side-rendering/#nextjs) ssr:true so I assumed that it is true by default.
- Turns out for urql the `store-name` comes from `ctx.query.storename` whereas in serversideprops it is `ctx.params.id`

> @rohanrajpal I didn't look at your code-fragment well, you have to return the fetchOptions object from a function, then the Router will work correctly.

One last doubt I have is why did we use Router in a function? What was conceptually wrong here? As you might have inferred already, I'm pretty new to the web dev world. 

Anyhow, learning from this, gonna read the docs a little more in depth in the future.

**Reply by JoviDeCroock** [reply]
When we use it as a plain object then it will get called before React/Next has initialized on the client making next think we are calling it at the wrong time. The only moment we want it to look at the router is when we are actually calling it from our React-world

**Reply by rohanrajpal** [reply]
Ohh got it, thanks Jovi!
